src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -74,13 +117,388 @@ public class ConsensusManager
+                ChainedHeaderBlock chainedHeaderBlock = null;
+
+                lock (this.treeLock)
+                {
+                    chainedHeaderBlock = this.chainedHeaderTree.GetChainedHeaderBlock(blockHash);
+                }
+
+                if (chainedHeaderBlock == null)
+                {
+                    this.logger.LogTrace("Block hash '{0}' is not part of the tree.", blockHash);
+                    onBlockDownloadedCallback(null);
+                    continue;
+                }
+
+                if (chainedHeaderBlock.Block != null)
+                {
+                    this.logger.LogTrace("Block pair '{0}' was found in memory.", chainedHeaderBlock);
+                    onBlockDownloadedCallback(chainedHeaderBlock);
+                    continue;
+                }
+
+                if (this.blockStore != null)
+                {
+                    Block block = await this.blockStore.GetBlockAsync(blockHash);
+                    if (block != null)
+                    {
+                        var newBlockPair = new ChainedHeaderBlock(block, chainedHeaderBlock.ChainedHeader);
+                        this.logger.LogTrace("Chained header block '{0}' was found in store.", newBlockPair);
+                        onBlockDownloadedCallback(newBlockPair);
+                        continue;
+                    }
+                }
+
+                // If the list of blocks hashes requested for download are not consecutive we batch them
+                if (previousBlock != null && chainedHeaderBlock.ChainedHeader.Previous.HashBlock != previousBlock.ChainedHeader.HashBlock)
+                {
+                    this.logger.LogTrace("Block header '{0}' is not consecutive to previous block header '{1}'.", chainedHeaderBlock, previousBlock);
+                    this.DownloadBlocks(downloadRequests, this.ProcessDownloadedBlock);
+                    downloadRequests = new List<uint256>();
+                }
+
+                previousBlock = chainedHeaderBlock;
+                downloadRequests.Add(blockHash);
+                this.logger.LogTrace("Block hash '{0}' is queued for download.", blockHash);
+            }
+
+            this.DownloadBlocks(downloadRequests, this.ProcessDownloadedBlock);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Processes items in <see cref="toDownloadQueue"/> and ask the block puller for blocks to download.
+        /// If the tree has too many unconsumed blocks or the download will wait, the amount of blocks to downloaded depends on the avg value in <see cref="IBlockPuller.AverageBlockSize"/>.
+        /// </summary>
+        /// <remarks>
+        /// Requests that have too many blocks will be split in batches.
+        /// </remarks>
+        private void ProcessDownloadQueueLocked()
+        {
+            this.logger.LogTrace("()");
+
+            const long MaxUnconsumedBlocksDataBytes = 1048576 * 10; // 10 mb
```

Comment:
Make it class constant with comment, But also 10 MB seems to be very small number, for bitcoin it could mean just six blocks being downloaded in parallel